File: scripts/algorithms/C/Check if a Parentheses String Can Be Valid/Check if a Parentheses String Can Be Valid.py
# Runtime: 460 ms (Top 26.24%) | Memory: 15.4 MB (Top 86.61%)
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        def validate(s: str, locked: str, op: str) -> bool:
            bal, wild = 0, 0
            for i in range(len(s)):
                if locked[i] == "1":
                    bal += 1 if s[i] == op else -1
                else:
                    wild += 1
                if wild + bal < 0:
                    return False
            return bal <= wild
        return len(s) % 2 == 0 and validate(s, locked, '(') and validate(s[::-1], locked[::-1], ')')

File: scripts/algorithms/C/Check if a Parentheses String Can Be Valid/Check if a Parentheses String Can Be Valid.cpp
// Runtime: 117 ms (Top 72.95%) | Memory: 27.1 MB (Top 80.99%)
class Solution {
public:
    bool canBeValid(string s, string locked) {

        int n = s.size();
        if(n&1) return false;

        int balance = 0;

        // First check balance from left to right
        // For opening '(' brackets

        for(int i=0; i<n; i++) {

            // If either char is ( or it is unlocked,
            // We can increment balance

            if(locked[i] == '0' || s[i] == '(') balance++;
            else balance--; // otherwise decrement balance, since it is ) and also locked

            // Since balance is negative, we have more ')'.
            // And there is no unlocked char available
            // SO, it is invalid string for sure

            if(balance < 0) return false;
        }

        // reset balance
        balance = 0;

        // Then also check balance from right to left
        // For closing ')' brackets

        for(int i=n-1; i>=0; i--) {

            // If either char is ) or it is unlocked,
            // We can increment balance

            if(locked[i] == '0' || s[i] == ')') balance++;
            else balance--;

            // Since balance is negative, we have more '('.
            // And there is no unlocked char available
            // SO, it is invalid string for sure

            if(balance < 0) return false;
        }

        return true;
    }
};